### PR TITLE
[#4372] Fixing of legacy tests failing locally

### LINKED ISSUE #4372: PylonsTestCase/flask-related test errors on master
Test errors on master due to dependencies between the tests, that circleci isn't seeing (because it runs them in parallel):
```
$ nosetests --ckan --with-pylons=test-core.ini ckan
No handlers could be found for logger "ckan.lib.uploader"
......................................./usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py:2181: SAWarning: Usage of the 'related attribute set' operation is not currently supported within the execution stage of the flush process. Results may not be consistent.  Consider using alternative event listeners or connection-level operations instead.
  % method)
.........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................../usr/lib/ckan/default/local/lib/python2.7/site-packages/sqlalchemy/sql/compiler.py:624: SAWarning: Can't resolve label reference 'total DESC'; converting to text() (this warning may be suppressed after 10 occurrences)
  util.ellipses_string(element.element))
.....................................................................................................................................................................................................S...........S................................................................................................................................................................................................................................................................................................S............................................................................................................................................................Error: you need to specify the user name.
............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................FFF.E.....................
======================================================================
ERROR: ckan.tests.test_common.TestCommonRequest.test_params_also_works_on_flask_request
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/vagrant/src/ckan/ckan/tests/test_common.py", line 227, in test_params_also_works_on_flask_request
    assert u'a' in ckan_request.args
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/werkzeug/local.py", line 347, in __getattr__
    return getattr(self._get_current_object(), name)
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/paste/registry.py", line 137, in __getattr__
    return getattr(self._current_obj(), attr)
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/webob/request.py", line 1181, in __getattr__
    raise AttributeError(attr)
AttributeError: args

======================================================================
FAIL: ckan.tests.test_common.TestCommonG.test_accessing_missing_key_raises_error_on_flask_request
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/vagrant/src/ckan/ckan/tests/test_common.py", line 273, in test_accessing_missing_key_raises_error_on_flask_request
    assert_raises(AttributeError, getattr, ckan_g, u'user')
AssertionError: AttributeError not raised

======================================================================
FAIL: ckan.tests.test_common.TestCommonG.test_can_also_use_c_on_a_flask_request
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/vagrant/src/ckan/ckan/tests/test_common.py", line 261, in test_can_also_use_c_on_a_flask_request
    eq_(ckan_c.user, u'example')
AssertionError: u'test.ckan.net' != u'example'

======================================================================
FAIL: ckan.tests.test_common.TestCommonG.test_flask_g_is_used_on_a_flask_request
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/vagrant/src/ckan/ckan/tests/test_common.py", line 247, in test_flask_g_is_used_on_a_flask_request
    assert u'flask.g' in text_type(ckan_g)
AssertionError

----------------------------------------------------------------------
Ran 1975 tests in 1090.621s

FAILED (SKIP=3, errors=1, failures=3)
```
The problem seems to be some bad clean-up after using PylonsTestCase that has been exposed for some reason following recent changes. I can reproduce the error quickly like this:
```
$ nosetests --ckan --with-pylons=test-core.ini ckan/tests/legacy/functional/api/test_user.py:TestCreateUserApiDisabled.test_user_create_api_disabled_anon ckan/tests/test_common.py:TestCommonG.test_accessing_missing_key_raises_error_on_flask_request -v -x
No handlers could be found for logger "ckan.lib.uploader"

======================================================================
FAIL: ckan.tests.test_common.TestCommonG.test_accessing_missing_key_raises_error_on_flask_request
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/ckan/default/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/vagrant/src/ckan/ckan/tests/test_common.py", line 273, in test_accessing_missing_key_raises_error_on_flask_request
    assert_raises(AttributeError, getattr, ckan_g, u'user')
AssertionError: AttributeError not raised

----------------------------------------------------------------------
Ran 2 tests in 2.451s

FAILED (failures=1)
```

### PR DESCRIPTION
**Pull request by tino097**
Fixes #4372

### Proposed fixes:
Changed all occurrences of `app=paste.fixture.TestApp(wsgi)` to use `helpers._get_test_app()
Removed using the `ckan.tests.legacy.pylons_controller`.

The tests @davidread  has mentioned in the issue, were failing because the '`ckan_g` was acting as pylons request 
```
assert u'flask.g' in text_type(ckan_g)
(Pdb) ckan_g
<pylons.util.AttribSafeContextObj at 0x7f7f31c629d0 user=u'test.ckan.net', userobj=<User id=2ba08fe3-798c-4c43-9a52-dc039b9587f3 name=test.ckan...tive>>
```
There are few commented parts and files, for which i would like to hear opinion if they could be removed 

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

File: ckan/tests/legacy/functional/api/test_email_notifications.py [hunk 1]
```diff
@@ -409,23 +399,17 @@ def test_01_no_emails_sent_if_turned_off(self):
...
         mock_mail_server.SmtpServerHarness,
-        pylons_controller.PylonsTestCase):
+        ControllerTestCase):
     '''Tests for the ckan.email_notifications_since config setting.'''
 
     @classmethod
     def setup_class(cls):
-        cls._original_config = config.copy()
 
         # Don't send email notifications for activities older than 1
```
Comment (davidread) [inline ckan/tests/legacy/functional/api/test_email_notifications.py]: I think this comment needs to be moved down a few lines to be with the corresponding code

**Comment by tino097** [comment]
Actually `'test_auth` wouldn't be removed, but i had problems to set the config values. There are lots of tests and these configs should be set by `helpers.change_config`

```
 config['ckan.auth.anon_create_dataset'] = False
 config['ckan.auth.create_dataset_if_not_in_organization'] = False
 config['ckan.auth.user_create_groups'] = False
 config['ckan.auth.user_create_organizations'] = False
 config['ckan.auth.user_delete_groups'] = False
 config['ckan.auth.user_delete_organizations'] = False
 config['ckan.auth.create_unowned_dataset'] = False
 config['ckan.auth.create_user_via_api'] = False
 config['ckan.auth.create_user_via_web'] = True
 config['ckan.auth.roles_that_cascade_to_sub_groups'] = 'admin'
```
on the corresponding test. I dont have other idea how to get over this

**Comment by amercader** [comment]
If the legacy `test_i18n.py` is causing trouble due to its usage of `PylonsTestCase` I think it's safe to remove it considering we are testing i18n on the [new tests](https://github.com/ckan/ckan/blob/master/ckan/tests/lib/test_i18n.py#L146) as well.

`test_auth.py` is another matter though. It's a pretty critical part of the tests so we should keep it or make an audit to make sure all tests are covered in the new ones. @tino097 the following diff (agaiinst master) will make the necessary changes to avoid using old Pylons test classes and dodgy app generators:

```diff
diff --git a/ckan/tests/legacy/logic/test_auth.py b/ckan/tests/legacy/logic/test_auth.py
index 0d3b7ba..9f74105 100644
--- a/ckan/tests/legacy/logic/test_auth.py
+++ b/ckan/tests/legacy/logic/test_auth.py
@@ -10,17 +10,38 @@ import ckan.authz as authz
 from ckan.lib.create_test_data import CreateTestData
 import json
 
+from ckan.tests.helpers import FunctionalTestBase, reset_db
+
+
+class TestAuth(FunctionalTestBase):
 
-class TestAuth(tests.WsgiAppCase):
     @classmethod
     def setup_class(cls):
+
+        super(TestAuth, cls).setup_class()
+
         admin_api = get_action('get_site_user')(
             {'model': model, 'ignore_auth': True}, {})['apikey']
         ## This is a mutable dict on the class level so tests can
         ## add apikeys as they go along
         cls.apikeys = {'sysadmin': str(admin_api), 'random_key': 'moo'}
 
-        cls._original_config = config.copy()
+        cls.app = cls._get_test_app()
+
+    @classmethod
+    def teardown_class(cls):
+
+        super(TestAuth, cls).teardown_class()
+
+        reset_db()
+
+    def setup(self):
+        # Override default setup as these tests don't work if we reset the
+        # database after each test
+        pass
+
+    @classmethod
+    def _apply_config_changes(cls, config):
 
         config['ckan.auth.anon_create_dataset'] = False
         config['ckan.auth.create_dataset_if_not_in_organization'] = False
@@ -33,17 +54,6 @@ class TestAuth(tests.WsgiAppCase):
         config['ckan.auth.create_user_via_web'] = True
         config['ckan.auth.roles_that_cascade_to_sub_groups'] = 'admin'
 
-        wsgiapp = ckan.config.middleware.make_app(
-            config['global_conf'], **config)
-        cls.app = paste.fixture.TestApp(wsgiapp)
-
-    @classmethod
-    def teardown_class(cls):
-        config.clear()
-        config.update(cls._original_config)
-
-        model.repo.rebuild_db()
-
     @classmethod
     def _call_api(cls, action, data, user, status=None):
         params = '%s=1' % json.dumps(data)
@@ -51,7 +61,7 @@ class TestAuth(tests.WsgiAppCase):
                             params=params,
                             extra_environ={'Authorization': cls.apikeys[user]},
                             status=[200, 403, 409])
-        if res.status != (status or 200):
+        if res.status_int != (status or 200):
             error = json.loads(res.body)['error']
             raise AssertionError('Status was %s but should be %s. Error: %s' %
                                  (res.status, status, error))
@@ -244,7 +254,9 @@ class TestAuthOrgHierarchy(TestAuth):
 
     @classmethod
     def setup_class(cls):
-#        TestAuth.setup_class()
+
+        super(TestAuth, cls).setup_class()
+
         CreateTestData.create_group_hierarchy_test_data()
 
         cls.apikeys = {}
@@ -254,23 +266,17 @@ class TestAuthOrgHierarchy(TestAuth):
         cls.sysadmin = get_action('get_site_user')(
             {'model': model, 'ignore_auth': True}, {})
 
-        cls._original_config = config.copy()
-
-        config['ckan.auth.roles_that_cascade_to_sub_groups'] = 'admin'
-
-        wsgiapp = ckan.config.middleware.make_app(
-            config['global_conf'], **config)
-        cls.app = paste.fixture.TestApp(wsgiapp)
-
         CreateTestData.create_arbitrary(
             package_dicts= [{'name': 'adataset',
                              'groups': ['national-health-service']}],
             extra_user_names=['john'])
 
+        cls.app = cls._get_test_app()
+
     @classmethod
-    def teardown_class(cls):
-        config.clear()
-        config.update(cls._original_config)
+    def _apply_config_changes(cls, config):
+
+        config['ckan.auth.roles_that_cascade_to_sub_groups'] = 'admin'
 
     def _reset_a_datasets_owner_org(self):
         rev = model.repo.new_revision()


```

**Comment by amercader** [comment]
Thanks @tino097, that was tedious but really necessary

**Comment by tino097** [comment]
With this PR, fixing all reported issues regarding tests 
#4088 #4087

**Comment by davidread** [comment]
Nice one!

**Pull request merged**